internal/tlsprofile/get.go

```diff
@@ -0,0 +1,45 @@
+package tlsprofile
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	openshiftconfigv1 "github.com/openshift/api/config/v1"
+
+	configv1alpha1 "github.com/os-observability/tempo-operator/apis/config/v1alpha1"
+)
+
+// ErrGetProfileFromCluster return when failed to get the cluster security policy in openshift.
+var ErrGetProfileFromCluster = errors.New("failed to get profile from cluster, using default TLS profile")
+
+// Get the profile according to the features configuration, if the policy is invalid or is not specified (empty string) this
+// should return the default profile (Intermediate), if openshift.ClusterTLSPolicy is enabled, it should get the profile
+// from the cluster, if the cluster return a unknow profile this should return an error.
+func Get(ctx context.Context, fg configv1alpha1.FeatureGates, c k8getter, log logr.Logger) (TLSProfileOptions, error) {
+	var tlsProfileType openshiftconfigv1.TLSSecurityProfile
+	var err error
+	var returnedErr error
+	// If ClusterTLSPolicy is enabled get the policy from the cluster
+	if fg.OpenShift.ClusterTLSPolicy {
+		tlsProfileType, err = getTLSProfileFromCluster(ctx, c)
+		if err != nil {
+			returnedErr = ErrGetProfileFromCluster
+			tlsProfileType = getDefaultTLSSecurityProfile()
+		}
+	} else {
+		tlsProfileType, err = getTLSSecurityProfile(configv1alpha1.TLSProfileType(fg.TLSProfile))
+		if err != nil {
```

Comment:
If I'm an admin and by mistake specify `featureGates.tlsProfile: modern` (note the lower case `modern`), then the operator will log an error but still use the default `Intermediate` profile?

I'd suggest to always return an error in this case, and at some other place in the code set the default value for the `featureGates.tlsProfile` setting. afaics currently we don't have a way to set default values for the `ProjectConfig` struct, I'd suggest we create a new function to set the default values for the `ProjectConfig` struct (similar to the Defaulter webhook) and call it in our `main.go` files directly after parsing the yaml file.

Reply:
I agree we should have a centralized way of setting defaults I think that could be done in a separate PR.

Reply:
On the other hand what I don't like about returning an error here and allow other part of the code to handle it is that this function encapsulated all of things related to tlsprofile, if I do in other way I'll lose that encapsulation and I would need to add that logic to the controller or other piece of code.

Reply:
I think what we should do in the future, (not in the PR) is to validate those inputs in some place, making sure we cannot allow invalid settings, and also set the defaults as you said. then I can get the value and use here. and we should not reach the error block for this funciton.

Reply:
> On the other hand what I don't like about returning an error here and allow other part of the code to handle it is that this function encapsulated all of things related to tlsprofile, if I do in other way I'll lose that encapsulation and I would need to add that logic to the controller or other piece of code.

I think any code must be able to return errors, I'd do it the same way as with the `ErrGetProfileFromCluster` error - return an error and let the error handling code in `microservices_controller.go` present the error to the user (via the status conditions).

> I think what we should do in the future, (not in the PR) is to validate those inputs in some place, making sure we cannot allow invalid settings (the operator can fail to start if that happens or log an error and set a default), and also set the defaults as you said. if that happens I can get the value and use here, we should not reach the error block for this function, because all values have been validated. Make sense?

Yep, setting defaults and validating the ProjectConfig in a separate PR sounds good to me.

Reply:
I've created https://github.com/os-observability/tempo-operator/issues/266 to track the defaulter/validator for the ProjectConfig.